---
title: "scNet"
author: "Taiyun Kim"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    # toc_float:
    #   collapsed: false
    #   smooth_scroll: true
    code_folding: show
    number_sections: true
vignette: >
  %\VignetteIndexEntry{scNet}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

<h1>Introduction</h1> 
This R package is a collection of single cell RNA-sequencing (scRNA-seq) data analysis functions developed by team of <a href = "http://www.maths.usyd.edu.au/u/SMS/bioinformatics/">**Sydney Precision Bioinformatics Research Group**</a> at The University of Sydney.

This package contains useful functions for analysis of scRNA-seq data including clustering, cell type identification, etc.


# Clustering

## Load Data


```{r}
# load library
suppressPackageStartupMessages({
  library(scNet)
  library(mclust)
})

# load dataset
data("GSE82187.sample")
dat <- GSE82187

dat <- log2(dat+1)

# show the data
dat[1:5,1:5]

# set number of clusters (classes defined in colnames)
nCs <- length(table(colnames(dat)))

# cell types
cellTypes = colnames(dat)

dat.processed = dat
```


## scClust

### Using scClust

#### Using `SIMLR` clustering using Pearson correlation

```{r warning=F, error=F}
# since we have already filtered genes, we will set 'geneFilter=1'. Alternatively, you can log transform your data without filtering and set 'geneFilter=0.8'. This will return same result.
dat.scClust <- dat.processed
colnames(dat.scClust) <- cellTypes
simlr.result <- scClust(dat.scClust, nCs, method = "simlr", similarity = "pearson", geneFilter = 1, seed = 1, cores.ratio = 0)

# Evaluate
adjustedRandIndex(cellTypes, simlr.result$y$cluster)
```

#### Using k-means clustering from `amap` package

```{r warning=F, error=F}
km.result <- scClust(dat.scClust, nCs, method = "kmeans", similarity = "pearson", geneFilter = 1, seed = 1)

# Evaluate
adjustedRandIndex(cellTypes, km.result$cluster)
```



### Benchmarking different metrics with `scClustBench`

#### Using `SIMLR` clustering to benchmark

```{r warning=F, error=F}
simlr.bench <- scClustBench(dat.scClust, nCs, method = "simlr", similarity = c("euclidean", "pearson"), cores = 1, rep = 1, cores.ratio = 0)

simlr.bench.eval <- evalScClustBench(simlr.bench, method = "simlr")

p <- plotSimlrEval(simlr.bench.eval)
p
```

#### Using k-means from `amap` package to benchamrk

```{r warning=F, error=F}
km.bench <- scClustBench(dat.scClust, nCs, method = "kmeans", similarity = c("euclidean", "pearson"), rep = 1, cores = 1)

km.bench.eval <- evalScClustBench(km.bench, method = "kmeans")

p <- plotKmeansEval(km.bench.eval)
p
```


# Post hoc cell type classification

## Load Data

```{r}
# load dataset
data(GSE87795_liver.development.data)
dat <- GSE87795_liver.development.data$data
cellTypes <- GSE87795_liver.development.data$cellTypes

# Dimensions of the dataset
dim(dat)

# Show dataset
dat[1:5,1:5]

# Cell types (pre-defined)
table(cellTypes)

# number of clusters
nCs <- length(table(cellTypes))
```

## Preprocessing

For single cell data, we filter genes with more than 80% of zeros across the dataset. Then we log2 transform the expression data to normalisation. Here, our dataset has already been processed.

```{r}
# # Filter low expressed genes
# del <- which(rowSums(dat == 0) / ncol(dat) >= 0.8)
# dat.filtered <- dat[-del,]
# 
# # log2 transformation
# dat.processed <- log2(dat.filtered + 1)
dat.processed <- dat
```


## Step-by-step guide to using scReClassify (mislabelled cell type correction)

### Dimension reduction

```{r}
# PCA procedure
dat.pc <- matPCs(dat.processed, 0.7)
dim(dat.pc)
```

### Run scReClassify

```{r}
# run scReClassify
cellTypes.reclassify <- multiAdaSampling(dat.pc, cellTypes, seed = 1, classifier = "svm", percent = 1, L = 10)

```

### Verification of cell type correction with marker gene

```{r}
# Verification by marker genes
End <- c("KDR", "LYVE1")
Meg <- c("ITGA2B", "ITGB3")
Mes <- c("MEST", "MMP2")
Ery <- c("HBA-A1", "HBB-BT")

# check examples
idx <- which(cellTypes.reclassify$final != cellTypes)
library(dplyr)
cbind(original=cellTypes[idx], reclassify=cellTypes.reclassify$final[idx]) %>%
  DT::datatable()

c1 <- dat.processed[, which(cellTypes=="Endothelial Cell")]
c2 <- dat.processed[, which(cellTypes=="Erythrocyte")]
c3 <- dat.processed[, which(cellTypes=="Hepatoblast")]
c4 <- dat.processed[, which(cellTypes=="Macrophage")]
c5 <- dat.processed[, which(cellTypes=="Megakaryocyte")]
c6 <- dat.processed[, which(cellTypes=="Mesenchymal Cell")]
cs <- rainbow(length(table(cellTypes)))

# (example 1 E13.5_C20)
#####
par(mfrow=c(1,2))
marker <- End[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(1, dat.processed[marker, which(colnames(dat.processed) %in% "E13.5_C20")], pch=16, col="red", cex=2)
marker <- End[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(1, dat.processed[marker, which(colnames(dat.processed) %in% "E13.5_C20")], pch=16, col="red", cex=2)
#####

# (example 2 E13.5_C14)
#####
par(mfrow=c(1,2))
marker <- Meg[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(5, dat.processed[marker, which(colnames(dat.processed) %in% "E13.5_C14")], pch=16, col="red", cex=2)
marker <- Meg[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(5, dat.processed[marker, which(colnames(dat.processed) %in% "E13.5_C14")], pch=16, col="red", cex=2)
#####

# (example 3 E16.5_2_C65)
#####
par(mfrow=c(1,2))
marker <- Mes[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(6, dat.processed[marker, which(colnames(dat.processed) %in% "E16.5_2_C65")], pch=16, col="red", cex=2)
marker <- Mes[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(6, dat.processed[marker, which(colnames(dat.processed) %in% "E16.5_2_C65")], pch=16, col="red", cex=2)
######

# (example 4 E16.5_C17)
#####
par(mfrow=c(1,2))
marker <- Mes[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(6, dat.processed[marker, which(colnames(dat.processed) %in% "E16.5_C17")], pch=16, col="red", cex=2)
marker <- Mes[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(6, dat.processed[marker, which(colnames(dat.processed) %in% "E16.5_C17")], pch=16, col="red", cex=2)
#####

# (example 5 E12.5_C72)
#####
par(mfrow=c(1,2))
marker <- Ery[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(2, dat.processed[marker, which(colnames(dat.processed) %in% "E12.5_C72")], pch=16, col="red", cex=2)
marker <- Ery[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(2, dat.processed[marker, which(colnames(dat.processed) %in% "E12.5_C72")], pch=16, col="red", cex=2)
#####

# (example 6 E12.5_C07)
#####
par(mfrow=c(1,2))
marker <- Ery[1]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(2, dat.processed[marker, which(colnames(dat.processed) %in% "E12.5_C07")], pch=16, col="red", cex=2)
marker <- Ery[2]
boxplot(c1[marker,], c2[marker,], c3[marker,], c4[marker,], c5[marker,], c6[marker,], col=cs, main=marker)
points(2, dat.processed[marker, which(colnames(dat.processed) %in% "E12.5_C07")], pch=16, col="red", cex=2)
#####

```


# SessionInfo

```{r}
sessionInfo()
```
